cli/replicaset/cmd/upgrade.go

```diff
@@ -0,0 +1,279 @@
+package replicasetcmd
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/tarantool/tt/cli/connector"
+	"github.com/tarantool/tt/cli/replicaset"
+	"github.com/tarantool/tt/cli/running"
+)
+
+var (
+	ChosenReplicasetAliases []string
+	LsnTimeout              int
+)
+
+type InstanceMeta struct {
+	run  running.InstanceCtx
+	conn connector.Connector
+}
+
+//go:embed lua/upgrade.lua
+var upgradeMasterLua string
+
+type SyncInfo struct {
+	LSN uint64  `mapstructure:"lsn"`
+	IID uint32  `mapstructure:"iid"`
+	Err *string `mapstructure:"err"`
+}
+
+// "FilterReplicasetsByAliases" filters the given replicaset list by chosen aliases
+// and returns the chosen replicasets. If a non-existent alias is found, it returns an error.
+func FilterReplicasetsByAliases(replicasets replicaset.Replicasets) ([]replicaset.Replicaset,
+	error) {
+	// If no aliases are provided, return all replicasets.
+	if len(ChosenReplicasetAliases) == 0 {
+		return replicasets.Replicasets, nil
+	}
+
+	// Create a map for fast lookup of replicasets by alias
+	replicasetMap := make(map[string]replicaset.Replicaset)
+	for _, rs := range replicasets.Replicasets {
+		replicasetMap[rs.Alias] = rs
+	}
+
+	var chosenReplicasets []replicaset.Replicaset
+	for _, alias := range ChosenReplicasetAliases {
+		rs, exists := replicasetMap[alias]
+		if !exists {
+			return nil, fmt.Errorf("replicaset with alias %q doesn't exist", alias)
+		}
+		chosenReplicasets = append(chosenReplicasets, rs)
+	}
+
+	return chosenReplicasets, nil
+}
+
+func Upgrade(discoveryCtx DiscoveryCtx) error {
+	replicasets, err := GetReplicasets(discoveryCtx)
+	if err != nil {
+		// This may be a single-instance application without Tarantool-3 config
+		// or instances.yml file.
+		if len(discoveryCtx.RunningCtx.Instances) == 1 {
```

Comment:
GetReplicasets should work in this case.

Reply:
Using the `GetReplicasets` command doesn't work with applications that lack a configuration. For example, when I start a single application:

```
/start-from-snap $ tree
.
├── init.lua
├── ready
├── tt.yaml
└── var
    ├── lib
    │   └── start-from-snap
    │       └── 00000000000000000000.snap
    ├── log
    │   └── start-from-snap
    │       └── tt.log
    └── run
        └── start-from-snap
            ├── tarantool.control
            ├── tarantool.sock
            └── tt.pid
```

Checking the application status with `tt status` gives the following output:

```
/start-from-snap $ tt status
 INSTANCE         STATUS   PID   MODE  CONFIG         BOX      UPSTREAM
 start-from-snap  RUNNING  3127  RW    uninitialized  running  --
```

However, when attempting to upgrade using `GetReplicasets`, I encounter an error:

```
/start-from-snap $ tt replicaset upgrade start-from-snap
   ⨯ config:get(): no instance config available yet
```